{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e93027d-50c3-497d-91b7-7bd6a07647f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        #########################################################\n",
      "        ###           __  __  ____  ______                    ### \n",
      "        ###          |  \\/  |/ __ \\|  ____/\\    _             ### \n",
      "        ###          | \\  / | |  | | |__ /  \\ _| |_           ### \n",
      "        ###          | |\\/| | |  | |  __/ /\\ \\_   _|          ###\n",
      "        ###          | |  | | |__| | | / ____ \\|_|            ###\n",
      "        ###          |_|  |_|\\____/|_|/_/    \\_\\              ###\n",
      "        ###                                                   ### \n",
      "        ######################################################### \n",
      "       \n",
      " \n",
      "        \n"
     ]
    }
   ],
   "source": [
    "from mofapy2.run.entry_point import entry_point\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import mofax as mfx\n",
    "import os\n",
    "\n",
    "# initialise the entry point\n",
    "ent = entry_point()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2cf02c5d-9aa2-45b5-ad88-d6c30a1e1a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_FOLDER = \"C:/Users/49152/Downloads/Multi-omics/SCOT_plus/AGW_results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "37a2e4cb-7c7d-42bb-bf3f-8e518da3b9a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>sample</th>\n",
       "      <th>value</th>\n",
       "      <th>view</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Eif2b4_prot</td>\n",
       "      <td>05J_C10_A10</td>\n",
       "      <td>-1.636681</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Slc38a2_prot</td>\n",
       "      <td>05J_C10_A10</td>\n",
       "      <td>-0.899462</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Wnk1_prot</td>\n",
       "      <td>05J_C10_A10</td>\n",
       "      <td>2.877517</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Krt6a_prot</td>\n",
       "      <td>05J_C10_A10</td>\n",
       "      <td>-1.962201</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rpl18_prot</td>\n",
       "      <td>05J_C10_A10</td>\n",
       "      <td>1.478620</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198985</th>\n",
       "      <td>Cstf2_rna</td>\n",
       "      <td>07J_SVEC_D9</td>\n",
       "      <td>-0.088757</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198986</th>\n",
       "      <td>Acsl1_rna</td>\n",
       "      <td>07J_SVEC_D9</td>\n",
       "      <td>-1.201857</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198987</th>\n",
       "      <td>Tlcd3a_rna</td>\n",
       "      <td>07J_SVEC_D9</td>\n",
       "      <td>-0.855548</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198988</th>\n",
       "      <td>Elp1_rna</td>\n",
       "      <td>07J_SVEC_D9</td>\n",
       "      <td>-0.036836</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198989</th>\n",
       "      <td>9130008F23Rik_rna</td>\n",
       "      <td>07J_SVEC_D9</td>\n",
       "      <td>-1.174741</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>198990 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  feature       sample     value             view\n",
       "0             Eif2b4_prot  05J_C10_A10 -1.636681       Proteomics\n",
       "1            Slc38a2_prot  05J_C10_A10 -0.899462       Proteomics\n",
       "2               Wnk1_prot  05J_C10_A10  2.877517       Proteomics\n",
       "3              Krt6a_prot  05J_C10_A10 -1.962201       Proteomics\n",
       "4              Rpl18_prot  05J_C10_A10  1.478620       Proteomics\n",
       "...                   ...          ...       ...              ...\n",
       "198985          Cstf2_rna  07J_SVEC_D9 -0.088757  Transcriptomics\n",
       "198986          Acsl1_rna  07J_SVEC_D9 -1.201857  Transcriptomics\n",
       "198987         Tlcd3a_rna  07J_SVEC_D9 -0.855548  Transcriptomics\n",
       "198988           Elp1_rna  07J_SVEC_D9 -0.036836  Transcriptomics\n",
       "198989  9130008F23Rik_rna  07J_SVEC_D9 -1.174741  Transcriptomics\n",
       "\n",
       "[198990 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# --- Read data ---\n",
    "proteomics = pd.read_csv(os.path.join(INPUT_FOLDER, \"Proteomics_view.csv\"))\n",
    "transcriptomics = pd.read_csv(os.path.join(INPUT_FOLDER, \"RNA_view.csv\"))\n",
    "\n",
    "# Move \"Gene\" column into index (like rownames in R)\n",
    "proteomics_mat = proteomics.set_index(\"Gene\")\n",
    "transcriptomics_mat = transcriptomics.set_index(\"Gene\")\n",
    "\n",
    "# --- Reshape to long format ---\n",
    "proteomics_long = (\n",
    "    proteomics_mat\n",
    "    .reset_index()\n",
    "    .melt(id_vars=\"Gene\", var_name=\"sample\", value_name=\"value\")\n",
    "    .rename(columns={\"Gene\": \"feature\"})\n",
    ")\n",
    "proteomics_long[\"view\"] = \"Proteomics\"\n",
    "# Make features unique by adding suffix\n",
    "proteomics_long[\"feature\"] = proteomics_long[\"feature\"].astype(str) + \"_prot\"\n",
    "\n",
    "transcriptomics_long = (\n",
    "    transcriptomics_mat\n",
    "    .reset_index()\n",
    "    .melt(id_vars=\"Gene\", var_name=\"sample\", value_name=\"value\")\n",
    "    .rename(columns={\"Gene\": \"feature\"})\n",
    ")\n",
    "transcriptomics_long[\"view\"] = \"Transcriptomics\"\n",
    "# Make features unique by adding suffix\n",
    "transcriptomics_long[\"feature\"] = transcriptomics_long[\"feature\"].astype(str) + \"_rna\"\n",
    "\n",
    "# --- Combine into one long DataFrame ---\n",
    "mofa_long = pd.concat([proteomics_long, transcriptomics_long], ignore_index=True)\n",
    "\n",
    "mofa_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "389b5a22-3b6c-48d3-b83f-41d7ea7455a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "No \"group\" column found in the data frame, we will assume a common group for all samples...\n",
      "\n",
      "\n",
      "Loaded group='single_group' view='Proteomics' with N=67 samples and D=990 features...\n",
      "Loaded group='single_group' view='Transcriptomics' with N=67 samples and D=1980 features...\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Data.frame format\n",
    "ent.set_data_df(mofa_long, likelihoods = [\"gaussian\",\"gaussian\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "69faeef6-3491-4c59-8b47-46ead3da4196",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model options:\n",
      "- Automatic Relevance Determination prior on the factors: False\n",
      "- Automatic Relevance Determination prior on the weights: True\n",
      "- Spike-and-slab prior on the factors: False\n",
      "- Spike-and-slab prior on the weights: True\n",
      "Likelihoods:\n",
      "- View 0 (Proteomics): gaussian\n",
      "- View 1 (Transcriptomics): gaussian\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ent.set_model_options(factors = 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4bc8a757-9921-4b46-83e2-9a89aee501ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GPU mode is activated\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ent.set_train_options(\n",
    "    convergence_mode = \"slow\", \n",
    "    dropR2 = 0, \n",
    "    gpu_mode = True, \n",
    "    seed = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "253fc5da-d208-41c9-a108-b29d20bb5d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "ent.build()\n",
    "\n",
    "ent.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a7ea470-7522-486a-9a98-5d96a08ec0b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the output\n",
    "ent.save(outfile=\"C:/Users/49152/Downloads/Multi-omics/SCOT_plus/MOFA_output/SCOTplus_trained_model.hdf5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6983e552-6695-417a-a7e8-3e68245ee206",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mofa_env_gpu",
   "language": "python",
   "name": "mofa_env_gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
