{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4c37a570-dbfe-47d9-a2a5-6c0db8e0fa78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        #########################################################\n",
      "        ###           __  __  ____  ______                    ### \n",
      "        ###          |  \\/  |/ __ \\|  ____/\\    _             ### \n",
      "        ###          | \\  / | |  | | |__ /  \\ _| |_           ### \n",
      "        ###          | |\\/| | |  | |  __/ /\\ \\_   _|          ###\n",
      "        ###          | |  | | |__| | | / ____ \\|_|            ###\n",
      "        ###          |_|  |_|\\____/|_|/_/    \\_\\              ###\n",
      "        ###                                                   ### \n",
      "        ######################################################### \n",
      "       \n",
      " \n",
      "        \n"
     ]
    }
   ],
   "source": [
    "from mofapy2.run.entry_point import entry_point\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import mofax as mfx\n",
    "\n",
    "# initialise the entry point\n",
    "ent = entry_point()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb1de3e1-b0e9-4e6e-a776-3065f22dedcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>sample</th>\n",
       "      <th>value</th>\n",
       "      <th>view</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Krt1_prot</td>\n",
       "      <td>C10_05J_bulk01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Slc38a2_prot</td>\n",
       "      <td>C10_05J_bulk01</td>\n",
       "      <td>-0.883269</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tnc_prot</td>\n",
       "      <td>C10_05J_bulk01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Camk1_prot</td>\n",
       "      <td>C10_05J_bulk01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Krt6a_prot</td>\n",
       "      <td>C10_05J_bulk01</td>\n",
       "      <td>-0.817864</td>\n",
       "      <td>Proteomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76995</th>\n",
       "      <td>Med11_rna</td>\n",
       "      <td>SVEC_07J_bulk13</td>\n",
       "      <td>0.750824</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76996</th>\n",
       "      <td>Riok1_rna</td>\n",
       "      <td>SVEC_07J_bulk13</td>\n",
       "      <td>0.966808</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76997</th>\n",
       "      <td>Bclaf1_rna</td>\n",
       "      <td>SVEC_07J_bulk13</td>\n",
       "      <td>0.726323</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76998</th>\n",
       "      <td>Rnaseh2a_rna</td>\n",
       "      <td>SVEC_07J_bulk13</td>\n",
       "      <td>0.596358</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76999</th>\n",
       "      <td>Kras_rna</td>\n",
       "      <td>SVEC_07J_bulk13</td>\n",
       "      <td>-1.950468</td>\n",
       "      <td>Transcriptomics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77000 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            feature           sample     value             view\n",
       "0         Krt1_prot   C10_05J_bulk01       NaN       Proteomics\n",
       "1      Slc38a2_prot   C10_05J_bulk01 -0.883269       Proteomics\n",
       "2          Tnc_prot   C10_05J_bulk01       NaN       Proteomics\n",
       "3        Camk1_prot   C10_05J_bulk01       NaN       Proteomics\n",
       "4        Krt6a_prot   C10_05J_bulk01 -0.817864       Proteomics\n",
       "...             ...              ...       ...              ...\n",
       "76995     Med11_rna  SVEC_07J_bulk13  0.750824  Transcriptomics\n",
       "76996     Riok1_rna  SVEC_07J_bulk13  0.966808  Transcriptomics\n",
       "76997    Bclaf1_rna  SVEC_07J_bulk13  0.726323  Transcriptomics\n",
       "76998  Rnaseh2a_rna  SVEC_07J_bulk13  0.596358  Transcriptomics\n",
       "76999      Kras_rna  SVEC_07J_bulk13 -1.950468  Transcriptomics\n",
       "\n",
       "[77000 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# --- Read data ---\n",
    "proteomics = pd.read_csv(\"C:/Users/49152/Downloads/Multi-omics/MOFA/input/2_Pseudobulk_different_composition/proteomics_pdbk_different_for_MOFA.csv\")\n",
    "transcriptomics = pd.read_csv(\"C:/Users/49152/Downloads/Multi-omics/MOFA/input/2_Pseudobulk_different_composition/transcriptomics_pdbk_different_for_MOFA.csv\")\n",
    "\n",
    "# Move \"Gene\" column into index (like rownames in R)\n",
    "proteomics_mat = proteomics.set_index(\"Gene\")\n",
    "transcriptomics_mat = transcriptomics.set_index(\"Gene\")\n",
    "\n",
    "# --- Reshape to long format ---\n",
    "proteomics_long = (\n",
    "    proteomics_mat\n",
    "    .reset_index()\n",
    "    .melt(id_vars=\"Gene\", var_name=\"sample\", value_name=\"value\")\n",
    "    .rename(columns={\"Gene\": \"feature\"})\n",
    ")\n",
    "proteomics_long[\"view\"] = \"Proteomics\"\n",
    "# Make features unique by adding suffix\n",
    "proteomics_long[\"feature\"] = proteomics_long[\"feature\"].astype(str) + \"_prot\"\n",
    "\n",
    "transcriptomics_long = (\n",
    "    transcriptomics_mat\n",
    "    .reset_index()\n",
    "    .melt(id_vars=\"Gene\", var_name=\"sample\", value_name=\"value\")\n",
    "    .rename(columns={\"Gene\": \"feature\"})\n",
    ")\n",
    "transcriptomics_long[\"view\"] = \"Transcriptomics\"\n",
    "# Make features unique by adding suffix\n",
    "transcriptomics_long[\"feature\"] = transcriptomics_long[\"feature\"].astype(str) + \"_rna\"\n",
    "\n",
    "# --- Combine into one long DataFrame ---\n",
    "mofa_long = pd.concat([proteomics_long, transcriptomics_long], ignore_index=True)\n",
    "\n",
    "mofa_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "929dea4b-e0c5-44aa-a4b1-91e15eef3110",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "No \"group\" column found in the data frame, we will assume a common group for all samples...\n",
      "\n",
      "\n",
      "Loaded group='single_group' view='Proteomics' with N=23 samples and D=1000 features...\n",
      "Loaded group='single_group' view='Transcriptomics' with N=27 samples and D=2000 features...\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Data.frame format\n",
    "ent.set_data_df(mofa_long, likelihoods = [\"gaussian\",\"gaussian\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "269f8c77-575d-479d-8874-f60cb8946d2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model options:\n",
      "- Automatic Relevance Determination prior on the factors: False\n",
      "- Automatic Relevance Determination prior on the weights: True\n",
      "- Spike-and-slab prior on the factors: False\n",
      "- Spike-and-slab prior on the weights: True\n",
      "Likelihoods:\n",
      "- View 0 (Proteomics): gaussian\n",
      "- View 1 (Transcriptomics): gaussian\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ent.set_model_options(factors = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "51da2bc3-ffba-4131-9d34-782fa49ce258",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GPU mode is activated\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ent.set_train_options(\n",
    "    convergence_mode = \"slow\", \n",
    "    dropR2 = 0.001, \n",
    "    gpu_mode = True, \n",
    "    seed = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e90a5b4-9f26-44c3-a548-9f361ff91d07",
   "metadata": {},
   "outputs": [],
   "source": [
    "ent.build()\n",
    "\n",
    "ent.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c23bb868-feef-4ac0-bec9-a4dbbfa15f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the output\n",
    "ent.save(outfile=\"C:/Users/49152/Downloads/Multi-omics/MOFA/output/pdbk_different_trained_model.hdf5\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mofa_env_gpu",
   "language": "python",
   "name": "mofa_env_gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
